spec/datadog/core/remote/component_spec.rb

```diff
@@ -0,0 +1,83 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+require 'datadog/core/remote/component'
+
+RSpec.describe Datadog::Core::Remote::Component do
+  let(:settings) { Datadog::Core::Configuration::Settings.new }
+  let(:agent_settings) { Datadog::Core::Configuration::AgentSettingsResolver.call(settings, logger: nil) }
+
+  describe '#initialize' do
+    subject(:component) { described_class.new(settings, agent_settings) }
+
+    let(:transport_v7) { double }
+    let(:client) { double }
+    let(:worker) { component.instance_eval { @worker } }
+
+    before do
+      expect(Datadog::Core::Transport::HTTP).to receive(:v7).and_return(transport_v7)
+      expect(Datadog::Core::Remote::Client).to receive(:new).and_return(client)
+
+      expect(worker).to receive(:start).and_call_original
+      expect(worker).to receive(:stop).and_call_original
+
+      component.barrier(:once)
+    end
+
+    after do
+      component.shutdown!
+    end
+
+    context 'when client sync succeeds' do
+      before do
+        expect(worker).to receive(:call).and_call_original
+        expect(client).to receive(:sync).and_return(nil)
+      end
+
+      it 'does not log any error' do
+        expect(Datadog.logger).to_not receive(:error)
+
+        Thread.pass # allow worker thread to work
+        sleep 0.1
+      end
+    end
+
+    context 'when client sync raises' do
+      let(:second_client) { double }
+
+      before do
+        expect(worker).to receive(:call).and_call_original
+        expect(client).to receive(:sync).and_raise(StandardError, 'test')
+      end
+
+      it 'logs an error' do
+        allow(Datadog::Core::Remote::Client).to receive(:new).and_return(client)
+
+        expect(Datadog.logger).to receive(:error).and_return(nil)
+
+        Thread.pass # allow worker thread to work
+        sleep 0.1
+      end
+
+      it 'catches exceptions' do
+        allow(Datadog::Core::Remote::Client).to receive(:new).and_return(client)
+
+        # if the error is uncaught it will crash the test, so a mere passing is good
+
+        Thread.pass # allow worker thread to work
+        sleep 0.1
+      end
+
+      it 'creates a new client' do
+        expect(Datadog::Core::Remote::Client).to receive(:new).and_return(second_client)
+
+        expect(component.client.object_id).to eql(client.object_id)
+
+        Thread.pass # allow worker thread to work
+        sleep 0.1
```

Comment:
We could do better and remove the thread pass and sleep. We could extract the code that passes the block to the worker on the component to a private method. In the test, we can override that method to still calls client sync, but use a queue to make sure the tests  waits until we know we are in the correct state.


```ruby
# remote/component.rb

Worker.new(interval: settings.remote.poll_interval_seconds) { sync_client }

private 

def sync_client
  @client.sync
end
```

```ruby
it 'creates a new client' do
  queue = Queue.new
  expect(Datadog::Core::Remote::Client).to receive(:new).and_return(second_client)
  expect(component.client.object_id).to eql(client.object_id)

  expect(component).to receive(:sync_client).and_wrap_original do |method, arg|
    method.call(arg)
    queue << 1
  end

  queue.pop
  expect(component.client.object_id).to eql(second_client.object_id)
end
```

Reply:
My plan is to replace the `Thread.pass` and `sleep` with an effectively working `barrier(:once)`, which will effectively wait until the worker has executed once.

Reply:
If you plan on replacing it with a `barrier(:once)` I would suggest to work on that on this PR

Reply:
Done